compiler/rustc_mir_transform/src/liveness.rs

```diff
@@ -0,0 +1,1073 @@
+    // First, report fully unused locals.
+    debug!("report fully unused places");
+    for (index, place) in checked_places.iter() {
+        if ever_live.contains(index) {
+            continue;
+        }
+
+        // This is a capture: let the enclosing function report the unused variable.
+        if is_capture(*place) {
+            debug_assert_eq!(place.local, ty::CAPTURE_STRUCT_LOCAL);
+            for p in place.projection {
+                if let PlaceElem::Field(f, _) = p {
+                    dead_captures.insert(*f);
+                    break;
+                }
+            }
+            continue;
+        }
+
+        let Some((ref name, def_span)) = checked_places.names[index] else { continue };
+        if name.is_empty() || name.starts_with('_') || name == "self" {
+            continue;
+        }
+
+        let local = place.local;
+        let decl = &body.local_decls[local];
+
+        if decl.from_compiler_desugaring() {
+            continue;
+        }
+
+        // Only report actual user-defined binding from now on.
+        let Some(LocalInfo::User(ClearCrossCrate::Set(local_info))) = decl.local_info.as_deref() else { continue };
+        let BindingForm::Var(binding) = local_info else { continue };
+        let Some(hir_id) = decl.source_info.scope.lint_root(&body.source_scopes) else { continue };
+
+        if let Some(ref mut assignments) = assignments.get_mut(&index) {
+            let introductions = &binding.introductions;
+
+            // Idiomatic rust assigns a value to a local upon definition.  However, we do not want to
+            // warn twice, for the unused local and for the unused assignment.  Therefore, we remove
+            // from the list of assignments the ones that happen at the definition site.
+            assignments.retain(|source_info, _| {
+                source_info.span.find_ancestor_inside(binding.pat_span).is_none()
+            });
+
+            // Extra assignments that we recognize thanks to the initialization span.  We need to
+            // take care of macro contexts here to be accurate.
+            if let Some((_, initializer_span)) = binding.opt_match_place {
+                assignments.retain(|source_info, _| {
+                    let within = source_info.span.find_ancestor_inside(initializer_span);
+                    let outer_initializer_span =
+                        initializer_span.find_ancestor_in_same_ctxt(source_info.span);
+                    within.is_none()
+                        && outer_initializer_span.map_or(true, |s| !s.contains(source_info.span))
+                });
+            }
+            debug!(?assignments);
+
+            if !assignments.is_empty() {
+                // We have a dead local with outstanding assignments and with non-trivial drop.
+                // This is probably a drop-guard, so we do not issue a warning there.
+                if ever_dropped.contains(index) {
+                    continue;
+                }
```

Comment:
Thus far this drop wasn't considered a use, since it makes the analysis very conservative. It wouldn't warn about strings, vecs, generic types without copy bound. For example:

```rust
pub fn f<T>(mut a: T, b: T) {
    a = b;
}
```

Reply:
My first version didn't consider drops either, but I ran into the issue of drop guards that have their fields assigned, and that we should not warn.
I can relax this check by trying to resolve the drop impl, and only consider drops that actually have a drop shim to be uses.

Reply:
It would be valuable to mention the rationale for the change in the pull request description. Without seeing those false positives it rather unclear why it is necessary now while it wasn't before. Not sure if I correctly surmise that the old implementation misclassified expressions with projections as a use, and drop guards generally need to be "released" with a field write, which hide the false positives.